File: src/agent.ts
import { createOpenRouter } from '@openrouter/ai-sdk-provider';
import { generateText, tool } from 'ai';
import { z } from 'zod';
import * as dotenv from 'dotenv';

dotenv.config();

const openrouter = createOpenRouter({
  apiKey: process.env.OPENROUTER_API_KEY || 'YOUR_OPENROUTER_API_KEY',
});

// Define tools for the agent
const tools = {
  calculator: tool({
    description: 'Perform basic arithmetic operations (add, subtract, multiply, divide)',
    parameters: z.object({
      operation: z.enum(['add', 'subtract', 'multiply', 'divide']).describe('The arithmetic operation to perform'),
      num1: z.number().describe('First number'),
      num2: z.number().describe('Second number'),
    }),
    execute: async (args) => {
      console.log(`  [Tool] calculator received args:`, JSON.stringify(args));
      const { operation, num1, num2 } = args;
      let result: number;
      switch (operation) {
        case 'add':
          result = num1 + num2;
          break;
        case 'subtract':
          result = num1 - num2;
          break;
        case 'multiply':
          result = num1 * num2;
          break;
        case 'divide':
          if (num2 === 0) throw new Error('Cannot divide by zero');
          result = num1 / num2;
          break;
        default:
          throw new Error(`Unknown operation: ${operation}`);
      }
      console.log(`  [Tool] calculator(${operation}, ${num1}, ${num2}) = ${result}`);
      return `The result of ${num1} ${operation} ${num2} is ${result}`;
    },
  }),

  getWeather: tool({
    description: 'Get the current weather for a city',
    parameters: z.object({
      city: z.string().describe('The city name'),
    }),
    execute: async ({ city }) => {
      // Simulate weather API call
      const weather = {
        temperature: Math.floor(Math.random() * 30) + 10,
        condition: ['sunny', 'cloudy', 'rainy', 'snowy'][Math.floor(Math.random() * 4)],
        humidity: Math.floor(Math.random() * 40) + 40,
      };
      console.log(`  [Tool] getWeather("${city}") = ${JSON.stringify(weather)}`);
      return `The weather in ${city} is ${weather.condition} with a temperature of ${weather.temperature}°C and ${weather.humidity}% humidity.`;
    },
  }),

  searchKnowledge: tool({
    description: 'Search a knowledge base for information',
    parameters: z.object({
      query: z.string().describe('The search query'),
    }),
    execute: async ({ query }) => {
      // Simulate knowledge base search
      const knowledge = {
        'typescript': 'TypeScript is a strongly typed programming language that builds on JavaScript.',
        'ai': 'Artificial Intelligence is the simulation of human intelligence by machines.',
        'nodejs': 'Node.js is a JavaScript runtime built on Chrome\'s V8 JavaScript engine.',
      };

      const lowerQuery = query.toLowerCase();
      for (const [key, value] of Object.entries(knowledge)) {
        if (lowerQuery.includes(key)) {
          console.log(`  [Tool] searchKnowledge("${query}") = Found match for "${key}"`);
          return value;
        }
      }

      console.log(`  [Tool] searchKnowledge("${query}") = No results found`);
      return 'No relevant information found in the knowledge base.';
    },
  }),

  getCurrentTime: tool({
    description: 'Get the current date and time',
    parameters: z.object({}),
    execute: async () => {
      const now = new Date();
      console.log(`  [Tool] getCurrentTime() = ${now.toLocaleString()}`);
      return now.toLocaleString();
    },
  }),
};

async function runAgent(userPrompt: string, maxSteps: number = 5) {
  console.log('='.repeat(80));
  console.log('AGENT EXECUTION');
  console.log('='.repeat(80));
  console.log(`User Prompt: ${userPrompt}`);
  console.log(`Max Steps: ${maxSteps}`);
  console.log('='.repeat(80));
  console.log('');

  const startTime = Date.now();

  try {
    const result = await generateText({
      model: openrouter.chat('minimax/minimax-m2'),
      prompt: userPrompt,
      tools: tools,
      maxSteps: maxSteps,
      onStepFinish: ({ text, toolCalls, toolResults, finishReason, usage }) => {
        console.log('\n--- Step Completed ---');
        if (text) {
          console.log(`Text: ${text.substring(0, 200)}${text.length > 200 ? '...' : ''}`);
        }
        if (toolCalls && toolCalls.length > 0) {
          console.log(`Tool Calls: ${toolCalls.length}`);
          toolCalls.forEach((call, index) => {
            console.log(`  ${index + 1}. ${call.toolName}`);
            if (call.args) {
              console.log(`     Args: ${JSON.stringify(call.args, null, 2)}`);
            } else {
              console.log(`     Args: (none)`);
            }
          });
        }
        if (toolResults && toolResults.length > 0) {
          console.log(`Tool Results: ${toolResults.length}`);
          toolResults.forEach((result, index) => {
            const resultStr = typeof result.result === 'string'
              ? result.result
              : JSON.stringify(result.result);
            const displayStr = resultStr ? resultStr.substring(0, 100) : '(empty)';
            console.log(`  ${index + 1}. ${result.toolName}: ${displayStr}`);
          });
        }
        console.log(`Finish Reason: ${finishReason || 'unknown'}`);
        console.log(`Tokens Used: ${usage?.totalTokens || 0}`);
        console.log('');
      },
    });

    const duration = Date.now() - startTime;

    console.log('='.repeat(80));
    console.log('FINAL RESULT');
    console.log('='.repeat(80));
    console.log(`\n${result.text || '(no text response)'}\n`);
    console.log('='.repeat(80));
    console.log(`Total Steps: ${result.steps?.length || 0}`);
    console.log(`Total Tokens: ${result.usage?.totalTokens || 0}`);
    console.log(`Duration: ${duration}ms`);
    console.log(`Finish Reason: ${result.finishReason || 'unknown'}`);
    console.log('='.repeat(80));
  } catch (error) {
    console.error('\n=== ERROR ===');
    console.error('Error during agent execution:');
    console.error(error);
    if (error instanceof Error) {
      console.error('Stack trace:', error.stack);
    }
    console.error('='.repeat(80));
  }
}

async function main() {
  try {
    // Example 1: Multi-step calculation
    await runAgent(
      'Calculate (15 + 25) * 3 and then divide the result by 10. Show me each step.',
      10
    );

    console.log('\n\n');

    // Example 2: Information retrieval and reasoning
    await runAgent(
      'What is TypeScript? After you find out, tell me the current time.',
      5
    );

    console.log('\n\n');

    // Example 3: Complex multi-tool task
    await runAgent(
      'Get the weather in London, then calculate what the temperature would be in Fahrenheit (use formula: F = C * 9/5 + 32).',
      10
    );

  } catch (error) {
    console.error('Error:', error instanceof Error ? error.message : String(error));
  }
}

main();

File: python/examples/agent.py
"""Agent with tool calling example.

Equivalent to TypeScript src/agent.ts and src/agent-simple.ts
Note: Includes debug logging to see actual responses from the model
"""

import sys
import os
import json
from datetime import datetime

# Add parent directory to path to import utils
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from utils.config import Config


# Define tool functions
def calculator(operation: str, num1: float, num2: float) -> str:
    """Perform basic arithmetic operations."""
    print(f"  [TOOL EXECUTED] calculator({operation}, {num1}, {num2})")

    if operation == "add":
        result = num1 + num2
    elif operation == "subtract":
        result = num1 - num2
    elif operation == "multiply":
        result = num1 * num2
    elif operation == "divide":
        if num2 == 0:
            return "Error: Cannot divide by zero"
        result = num1 / num2
    else:
        return f"Error: Unknown operation '{operation}'"

    print(f"  [TOOL RESULT] {num1} {operation} {num2} = {result}")
    return f"The result of {num1} {operation} {num2} is {result}"


def get_weather(city: str) -> str:
    """Get simulated weather for a city."""
    import random

    print(f"  [TOOL EXECUTED] get_weather('{city}')")

    weather_conditions = ["sunny", "cloudy", "rainy", "snowy"]
    weather = {
        "temperature": random.randint(10, 30),
        "condition": random.choice(weather_conditions),
        "humidity": random.randint(40, 80)
    }

    result = (f"The weather in {city} is {weather['condition']} "
              f"with a temperature of {weather['temperature']}°C "
              f"and {weather['humidity']}% humidity.")

    print(f"  [TOOL RESULT] {result}")
    return result


def get_current_time() -> str:
    """Get the current date and time."""
    print("  [TOOL EXECUTED] get_current_time()")

    now = datetime.now()
    result = now.strftime("%Y-%m-%d %H:%M:%S")

    print(f"  [TOOL RESULT] {result}")
    return f"Current time is {result}"


# Define tool schemas
tools = [
    {
        "type": "function",
        "function": {
            "name": "calculator",
            "description": "Perform basic arithmetic operations (add, subtract, multiply, divide)",
            "parameters": {
                "type": "object",
                "properties": {
                    "operation": {
                        "type": "string",
                        "enum": ["add", "subtract", "multiply", "divide"],
                        "description": "The arithmetic operation to perform"
                    },
                    "num1": {
                        "type": "number",
                        "description": "First number"
                    },
                    "num2": {
                        "type": "number",
                        "description": "Second number"
                    }
                },
                "required": ["operation", "num1", "num2"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": "Get the current weather for a city",
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {
                        "type": "string",
                        "description": "The city name, e.g., 'London' or 'New York'"
                    }
                },
                "required": ["city"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_current_time",
            "description": "Get the current date and time",
            "parameters": {
                "type": "object",
                "properties": {}
            }
        }
    }
]

# Map function names to actual functions
available_functions = {
    "calculator": calculator,
    "get_weather": get_weather,
    "get_current_time": get_current_time
}


def run_agent(user_message: str, max_iterations: int = 5):
    """Run agent with tool calling capabilities."""
    client = Config.get_client()

    print("=" * 80)
    print("AGENT EXECUTION")
    print("=" * 80)
    print(f"User Prompt: {user_message}")
    print(f"Max Iterations: {max_iterations}")
    print("=" * 80)
    print()

    messages = [{"role": "user", "content": user_message}]

    for iteration in range(max_iterations):
        print(f"\n{'='*80}")
        print(f"ITERATION {iteration + 1}")
        print(f"{'='*80}")

        # Make API call
        response = client.chat.completions.create(
            model=Config.MODEL_FREE,
            messages=messages,
            tools=tools
        )

        response_message = response.choices[0].message

        # FIRST: Check for thinking/reasoning in this iteration
        if hasattr(response_message, 'reasoning') and response_message.reasoning:
            reasoning_text = response_message.reasoning

            print(f"\n{'─'*80}")
            print("🧠 MODEL'S THINKING (this iteration):")
            print(f"{'─'*80}")

            # Extract and show thinking block
            if '<think>' in reasoning_text and '</think>' in reasoning_text:
                thinking = reasoning_text.split('<think>')[1].split('</think>')[0].strip()
                print(f"\n{thinking}\n")
                print(f"{'─'*80}")
            else:
                print(f"\n{reasoning_text}\n")
                print(f"{'─'*80}")

        # Debug: Print the full response
        print(f"\n📊 Response Metadata:")
        print(f"  finish_reason: {response.choices[0].finish_reason}")
        print(f"  has tool_calls: {response_message.tool_calls is not None}")
        print(f"  has content: {bool(response_message.content)}")
        print(f"  has reasoning: {hasattr(response_message, 'reasoning') and bool(response_message.reasoning)}")

        if response_message.tool_calls:
            print(f"\n🔧 Tool Calls: {len(response_message.tool_calls)}")

            # Add assistant's response to messages
            messages.append(response_message)

            # Execute each tool call
            for idx, tool_call in enumerate(response_message.tool_calls, 1):
                function_name = tool_call.function.name
                function_args = json.loads(tool_call.function.arguments)

                print(f"\n  Tool {idx}: {function_name}")
                print(f"  Arguments: {json.dumps(function_args, indent=2)}")

                # Call the function
                if function_name in available_functions:
                    function_response = available_functions[function_name](**function_args)
                else:
                    function_response = f"Error: Function '{function_name}' not found"

                # Add function response to messages
                messages.append({
                    "role": "tool",
                    "tool_call_id": tool_call.id,
                    "content": str(function_response)
                })

        else:
            # No more tool calls, return final response
            final_text = response_message.content

            # MiniMax M2 special handling: Check reasoning field if content is empty
            if not final_text and hasattr(response_message, 'reasoning') and response_message.reasoning:
                reasoning_text = response_message.reasoning

                print("\n" + "=" * 80)
                print("MODEL'S INTERNAL THINKING PROCESS")
                print("=" * 80)

                # Extract text after </think> tag if present
                if '</think>' in reasoning_text:
                    # Split into thinking and response
                    parts = reasoning_text.split('</think>')
                    thinking = parts[0].replace('<think>', '').strip()
                    final_text = parts[-1].strip()

                    print(f"\n{thinking}\n")
                    print("=" * 80)
                    print("END OF THINKING - MODEL'S FINAL RESPONSE BELOW")
                else:
                    final_text = reasoning_text
                    print(f"\n{reasoning_text}\n")

                print("=" * 80)

            print("\n" + "=" * 80)
            print("FINAL RESPONSE")
            print("=" * 80)
            print(f"\n{final_text}\n")
            print("=" * 80)
            print(f"Total Iterations: {iteration + 1}")
            print(f"Total Tokens: {response.usage.total_tokens}")
            print("=" * 80)

            return final_text

    print("\n" + "=" * 80)
    print("Max iterations reached without final response")
    print("=" * 80)
    return None


def main():
    """Run agent examples."""
    print("Testing MiniMax M2 Free Model - Agent with Tool Calling\n")
    print("This will show the model's internal thinking process!\n")

    try:
        # Example 1: Multi-step calculation
        run_agent("Calculate (15 + 25) * 2 and tell me if the result is greater than 75")

        print("\n\n")

        # Example 2: Weather with analysis
        run_agent("What's the weather in Paris? Based on the conditions, should I bring an umbrella?")

        print("\n\n")

        # Example 3: Simple query
        run_agent("What is 100 divided by 4?")

    except Exception as error:
        print(f"Error: {error}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    main()
